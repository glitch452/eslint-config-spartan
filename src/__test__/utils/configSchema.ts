import { z } from 'zod';

const initialEcmaVersion = 3;
const initialEcmaYear = 2015;
const nextYear = new Date().getFullYear() + 1;
const minEcmaVersion = 5;
const maxEcmaVersion = nextYear - initialEcmaYear + minEcmaVersion + 1;

const ecmaVersionSchema = z.union([
  z.literal('latest'),
  z.literal(initialEcmaVersion),
  z.number().int().min(minEcmaVersion).max(maxEcmaVersion),
  z.number().int().min(initialEcmaYear).max(nextYear),
]);
const sourceTypeSchema = z.enum(['script', 'module', 'commonjs']);
const globalConfSchema = z.boolean().or(z.enum(['off', 'readable', 'readonly', 'writable', 'writeable']));

const minSeverity = 0;
const maxSeverity = 2;
const severitySchema = z.number().int().min(minSeverity).max(maxSeverity);
const stringSeveritySchema = z.enum(['off', 'warn', 'error']);

const objectMetaPropertiesSchema = z.object({
  /** @deprecated Use `meta.name` instead. */
  name: z.string().optional(),
  /** @deprecated Use `meta.version` instead. */
  version: z.string().optional(),
  meta: z.object({ name: z.string().optional(), version: z.string().optional() }).optional(),
});

const esTreeParserSchema = z.union([
  objectMetaPropertiesSchema.extend({ parse: z.function() }),
  objectMetaPropertiesSchema.extend({ parseForESLint: z.function() }),
]);

const processorSchema = objectMetaPropertiesSchema.extend({
  supportsAutofix: z.boolean().optional(),
  preprocess: z.function(),
  postprocess: z.function(),
});

const parserOptionsSchema = z
  .object({
    ecmaVersion: ecmaVersionSchema,
    sourceType: sourceTypeSchema,
    ecmaFeatures: z
      .object({
        globalReturn: z.boolean(),
        impliedStrict: z.boolean(),
        jsx: z.boolean(),
        experimentalObjectRestSpread: z.boolean(),
      })
      .partial()
      .catchall(z.any()),
  })
  .partial()
  .catchall(z.any());

const languageOptionsSchema = z
  .object({
    ecmaVersion: ecmaVersionSchema,
    sourceType: sourceTypeSchema,
    globals: z.record(globalConfSchema),
    parser: esTreeParserSchema,
    parserOptions: parserOptionsSchema,
  })
  .partial();

const linterOptionsSchema = z
  .object({
    noInlineConfig: z.boolean().optional(),
    reportUnusedDisableDirectives: z.union([severitySchema, stringSeveritySchema, z.boolean()]).optional(),
  })
  .optional();

const ruleMetaDataSchema = z
  .object({
    docs: z
      .object({
        description: z.string(),
        category: z.string(),
        // Added z.string() to accommodate storybook plugin
        // Added z.record(z.unknown()) to accommodate typescript-eslint plugin
        recommended: z.union([z.boolean(), z.string(), z.record(z.unknown())]),
        url: z.string(),
      })
      .partial(),
    messages: z.record(z.string()),
    // Added z.null() added to accommodate the cypress plugin
    fixable: z.enum(['code', 'whitespace']).or(z.null()),
    schema: z.unknown(),
    deprecated: z.boolean(),
    replacedBy: z.string().array().readonly(),
    // Added 'error' to accommodate the security plugin
    type: z.enum(['problem', 'suggestion', 'layout', 'error']),
    hasSuggestions: z.boolean(),
  })
  .partial();

const ruleModuleSchema = z.object({
  create: z.function(),
  meta: ruleMetaDataSchema.optional(),
});

const pluginSchema = z
  .object({
    configs: z.record(z.unknown()),
    environments: z.record(
      z.object({ globals: z.record(globalConfSchema), parserOptions: parserOptionsSchema }).partial(),
    ),
    processors: z.record(processorSchema),
    rules: z.record(ruleModuleSchema),
  })
  .partial();

export const configSchema = z
  .object({
    name: z.string(),
    files: z.array(z.string().or(z.string().array())),
    ignores: z.string().array(),
    languageOptions: languageOptionsSchema,
    linterOptions: linterOptionsSchema,
    processor: z.string().or(processorSchema),
    plugins: z.record(pluginSchema),
    rules: z.unknown(),
    settings: z.record(z.unknown()),
  })
  .partial();
