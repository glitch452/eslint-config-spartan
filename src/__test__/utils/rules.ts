import { Linter, Rule } from 'eslint';
import { isOff, isWarn } from '../../utils/index.js';

/**
 * Extract a list of rule names that are 'enabled' i.e. set to 'warn' or 'error'
 * @param rules
 */
export function getEnabledRules(rules: Partial<Record<string, Linter.RuleEntry>> = {}): string[] {
  return Object.entries(rules)
    .filter(([, entry]) => entry && !isOff(entry))
    .map(([name]) => name);
}

/**
 * Extract a list of rule names that are set to 'warn'
 * @param rules
 */
export function getWarnRules(rules: Partial<Record<string, Linter.RuleEntry>> = {}): string[] {
  return Object.entries(rules)
    .filter(([, entry]) => entry && isWarn(entry))
    .map(([name]) => name);
}

/**
 * Extract a list of all the rule names with the given prefix prepended to them
 * @param rules
 * @param prefix
 * @param shouldSort
 */
export function listRules(
  rules:
    | Record<string, Linter.RuleEntry | Rule.RuleModule>
    | Map<string, Linter.RuleEntry | Rule.RuleModule>
    | undefined,
  prefix?: string,
  shouldSort: boolean = false,
): string[] {
  const keys = rules instanceof Map ? [...rules.keys()] : Object.keys(rules ?? {});
  const names = keys.map((x) => `${prefix ? `${prefix}/` : ''}${x}`);
  return shouldSort ? names.sort((a, b) => a.localeCompare(b)) : names;
}

/**
 * Extract a list of rule names that are deprecated and, if provided, prepend the prefix to the rule name
 * @param rules
 * @param prefix
 */
export function getDeprecatedRules(
  rules: Record<string, Rule.RuleModule> | Map<string, Rule.RuleModule> | undefined,
  prefix: string = '',
): string[] {
  const entries = rules instanceof Map ? [...rules.entries()] : Object.entries(rules ?? {});
  return entries.filter(([, rule]) => rule.meta?.deprecated).map(([name]) => `${prefix ? `${prefix}/` : ''}${name}`);
}

interface Docs extends NonNullable<Rule.RuleMetaData['docs']> {
  requiresTypeChecking?: boolean;
}

/**
 * Extract a list of rule names that require type information, if provided, prepend the prefix to the rule name
 * @param rules
 * @param prefix
 */
export function getTypeInfoRequiredRules(
  rules: Record<string, Rule.RuleModule> | Map<string, Rule.RuleModule> | undefined,
  prefix: string = '',
): string[] {
  const entries = rules instanceof Map ? [...rules.entries()] : Object.entries(rules ?? {});
  return (
    entries
      // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
      .filter(([, rule]) => (rule.meta?.docs as Docs).requiresTypeChecking)
      .map(([name]) => `${prefix ? `${prefix}/` : ''}${name}`)
  );
}
