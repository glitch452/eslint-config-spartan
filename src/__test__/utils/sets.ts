type Widen<T> = T | (TSReset.WidenLiteral<T> & {});

/**
 * `A âˆ© B`.
 * Compute the intersection of unique elements contained in both {@link a} and {@link b}.
 * @param a
 * @param b
 * @returns
 */
export function intersection<A, B>(a: Widen<A>[], b: Widen<B>[]): Extract<Widen<A>, Widen<B>>[] {
  const setB = new Set<Widen<A> | Widen<B>>(b);
  return [...new Set<Widen<A>>(a)].filter((x): x is Extract<Widen<A>, Widen<B>> => setB.has(x));
}

/**
 * `A - B` a.k.a. `A / B`.
 * Compute the unique elements in {@link a} that do not also exist in {@link b}.
 * @param a
 * @param b
 * @returns
 */
export function difference<A, B>(a: A[], b: B[]): Widen<A>[] {
  const setB = new Set<A | B>(b);
  return [...new Set<A>(a)].filter((x): x is A => !setB.has(x));
}
