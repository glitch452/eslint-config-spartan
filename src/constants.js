export const configNamePrefix = 'eslint-config-spartan';

export const prefixes = /** @type {const} */ ({
  chaiExpect: 'chai-expect',
  chaiFriendly: 'chai-friendly',
  cypress: 'cypress',
  i18next: 'i18next',
  i18nJson: 'i18n-json',
  import: 'import',
  jest: 'jest',
  jestDom: 'jest-dom',
  jsDoc: 'jsdoc',
  jsxA11y: 'jsx-a11y',
  mdx: 'mdx',
  mocha: 'mocha',
  next: '@next/next',
  promise: 'promise',
  react: 'react',
  reactHooks: 'react-hooks',
  regExp: 'regexp',
  security: 'security',
  storybook: 'storybook',
  stylistic: '@stylistic',
  tailwindCss: 'tailwindcss',
  testingLibrary: 'testing-library',
  typescriptEnum: 'typescript-enum',
  typescriptEsLint: '@typescript-eslint',
  unicorn: 'unicorn',
  unusedImports: 'unused-imports',
  vitest: 'vitest',
});
