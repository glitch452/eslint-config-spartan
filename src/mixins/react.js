import reactPlugin from 'eslint-plugin-react';
// eslint-disable-next-line import/default
import reactHooksPlugin from 'eslint-plugin-react-hooks';
import globals from 'globals';
import { fixupPluginRules } from '@eslint/compat';
import { files, warnToError } from '../utils/index.js';
import { CONFIG_NAME_PREFIX, prefixes } from '../constants.js';
/** @import { Linter } from 'eslint' */

/**
 * The `react` mixin creates an ESLint config for [eslint-plugin-react](https://www.npmjs.com/package/eslint-plugin-react)
 * and [eslint-plugin-react-hooks](https://www.npmjs.com/package/eslint-plugin-react-hooks) to assist with
 * [React](https://react.dev) application development.
 *
 * Rule Prefixes:
 *
 * - The `eslint-plugin-react` rules are prefixed with `react`
 * - The `eslint-plugin-react-hooks` rules are prefixed with `react-hooks`
 * @param {object} options
 * @param {Linter.Config['files']=} options.files Set the files for this config. By default, this applies to all js/ts variant file extensions.
 * @param {boolean=} options.noJsxRuntime Set this to true to disable the 'jsx-runtime' preset rules (JSX Runtime is available from React v17+).
 * @returns {Linter.Config}
 */
export function react(options = {}) {
  return {
    ...reactPlugin.configs.flat.recommended,
    name: `${CONFIG_NAME_PREFIX}/${react.name}`,
    files: options.files ?? [files.jsTs],
    languageOptions: {
      ...reactPlugin.configs.flat.recommended.languageOptions,
      globals: globals.browser,
    },
    settings: { react: { version: 'detect' } },
    plugins: {
      [prefixes.react]: reactPlugin,
      // Use fixupPluginRules for ESLint v9 support until the plugin is updated
      [prefixes.reactHooks]: fixupPluginRules(reactHooksPlugin),
    },
    rules: {
      ...reactPlugin.configs.flat.recommended.rules,
      ...(options.noJsxRuntime ? undefined : reactPlugin.configs['jsx-runtime'].rules),
      ...warnToError(reactHooksPlugin.configs.recommended.rules),
      'react/button-has-type': 'error',
      'react/hook-use-state': 'error',
      'react/jsx-boolean-value': ['error', 'never', { assumeUndefinedIsFalse: true }],
      'react/jsx-curly-brace-presence': ['error', 'never'],
      'react/jsx-fragments': 'error',
      'react/jsx-no-leaked-render': ['error', { validStrategies: ['coerce', 'ternary'] }],
      'react/jsx-no-useless-fragment': 'error',
      'react/jsx-pascal-case': ['error', { allowNamespace: true }],
      // 'react/jsx-sort-props': 'error', // Maybe?
      'react/no-array-index-key': 'error',
      // 'react/no-invalid-html-attribute': 'error', // Maybe?
      // 'react/prefer-read-only-props': 'error', // Maybe?
      'react/self-closing-comp': 'error',
      'react/void-dom-elements-no-children': 'error',
    },
  };
}
