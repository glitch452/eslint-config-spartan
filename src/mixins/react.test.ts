import reactPlugin from 'eslint-plugin-react';
import * as reactHooksPlugin from 'eslint-plugin-react-hooks';
import { react } from './react.js';
import { configSchema } from '../__test__/utils/configSchema.js';
import { difference, intersection } from '../__test__/utils/sets.js';
import { files } from '../utils/index.js';
import { CONFIG_NAME_PREFIX, prefixes } from '../constants.js';
import { getDeprecatedRules, getEnabledRules, getWarnRules, listRules } from '../__test__/utils/rules.js';
import globals from 'globals';

describe(react.name, () => {
  const config = react();

  const validRules = [
    ...listRules(reactPlugin.rules, prefixes.react),
    ...listRules(reactHooksPlugin.rules, prefixes.reactHooks),
  ];
  const deprecatedRules = [
    ...getDeprecatedRules(reactPlugin.rules, prefixes.react),
    ...getDeprecatedRules(reactHooksPlugin.rules, prefixes.reactHooks),
  ];

  it('should create a valid eslint config', () => {
    const actual = () => configSchema.parse(config);
    expect(actual).not.toThrow();
  });

  it('should add the expected plugins to the config', () => {
    const actual = Object.keys(config.plugins ?? {}).sort();
    const expected = [prefixes.react, prefixes.reactHooks].sort();
    expect(actual).toStrictEqual(expected);
  });

  it('should configure the expected rules when "noJsxRuntime" is not set', () => {
    expect(config.rules).toMatchSnapshot();
  });

  it('should configure the expected rules when "noJsxRuntime" is true', () => {
    const actual = react({ noJsxRuntime: true });
    expect(actual.rules).toMatchSnapshot();
  });

  it('should set the name value', () => {
    expect(config.name).toBe(`${CONFIG_NAME_PREFIX}/${react.name}`);
  });

  it('should set the default files value', () => {
    const actual = config.files;
    const expected = [files.jsTs];
    expect(actual).toStrictEqual(expected);
  });

  it('should set the given files value', () => {
    const fileValue = ['files.txt'];
    const actual = react({ files: fileValue }).files;
    expect(actual).toStrictEqual(fileValue);
  });

  it('should set the globals to the browser values', () => {
    const actual = config.languageOptions?.globals;
    const expected = globals.browser;
    expect(actual).toStrictEqual(expected);
  });

  it('should set the react.version setting to "detect"', () => {
    const actual = (config.settings as any)?.react?.version;
    const expected = 'detect';
    expect(actual).toStrictEqual(expected);
  });

  it('should only configure rules that exist', () => {
    const configuredRules = Object.keys(config.rules ?? {});
    const actual = difference(configuredRules, validRules);
    expect(actual).toStrictEqual([]);
  });

  it('should not configure rules that are deprecated', () => {
    const configuredAndEnabledRules = getEnabledRules(config.rules);
    const actual = intersection(configuredAndEnabledRules, deprecatedRules);
    expect(actual).toStrictEqual([]);
  });

  it('should not configure any rules as warn', () => {
    const configuredAndWarnRules = getWarnRules(config.rules);
    expect(configuredAndWarnRules).toStrictEqual([]);
  });
});
