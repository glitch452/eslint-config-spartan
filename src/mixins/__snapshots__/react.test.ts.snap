// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`react > should configure the expected rules when "noJsxRuntime" is not set 1`] = `
{
  "react-hooks/exhaustive-deps": "error",
  "react-hooks/rules-of-hooks": "error",
  "react/button-has-type": "error",
  "react/display-name": 2,
  "react/hook-use-state": "error",
  "react/jsx-boolean-value": [
    "error",
    "never",
    {
      "assumeUndefinedIsFalse": true,
    },
  ],
  "react/jsx-curly-brace-presence": [
    "error",
    "never",
  ],
  "react/jsx-fragments": "error",
  "react/jsx-key": 2,
  "react/jsx-no-comment-textnodes": 2,
  "react/jsx-no-duplicate-props": 2,
  "react/jsx-no-leaked-render": [
    "error",
    {
      "validStrategies": [
        "coerce",
        "ternary",
      ],
    },
  ],
  "react/jsx-no-target-blank": 2,
  "react/jsx-no-undef": 2,
  "react/jsx-no-useless-fragment": "error",
  "react/jsx-pascal-case": [
    "error",
    {
      "allowNamespace": true,
    },
  ],
  "react/jsx-uses-react": 0,
  "react/jsx-uses-vars": 2,
  "react/no-array-index-key": "error",
  "react/no-children-prop": 2,
  "react/no-danger-with-children": 2,
  "react/no-deprecated": 2,
  "react/no-direct-mutation-state": 2,
  "react/no-find-dom-node": 2,
  "react/no-is-mounted": 2,
  "react/no-render-return-value": 2,
  "react/no-string-refs": 2,
  "react/no-unescaped-entities": 2,
  "react/no-unknown-property": 2,
  "react/no-unsafe": 0,
  "react/prop-types": 2,
  "react/react-in-jsx-scope": 0,
  "react/require-render-return": 2,
  "react/self-closing-comp": "error",
  "react/void-dom-elements-no-children": "error",
}
`;

exports[`react > should configure the expected rules when "noJsxRuntime" is true 1`] = `
{
  "react-hooks/exhaustive-deps": "error",
  "react-hooks/rules-of-hooks": "error",
  "react/button-has-type": "error",
  "react/display-name": 2,
  "react/hook-use-state": "error",
  "react/jsx-boolean-value": [
    "error",
    "never",
    {
      "assumeUndefinedIsFalse": true,
    },
  ],
  "react/jsx-curly-brace-presence": [
    "error",
    "never",
  ],
  "react/jsx-fragments": "error",
  "react/jsx-key": 2,
  "react/jsx-no-comment-textnodes": 2,
  "react/jsx-no-duplicate-props": 2,
  "react/jsx-no-leaked-render": [
    "error",
    {
      "validStrategies": [
        "coerce",
        "ternary",
      ],
    },
  ],
  "react/jsx-no-target-blank": 2,
  "react/jsx-no-undef": 2,
  "react/jsx-no-useless-fragment": "error",
  "react/jsx-pascal-case": [
    "error",
    {
      "allowNamespace": true,
    },
  ],
  "react/jsx-uses-react": 2,
  "react/jsx-uses-vars": 2,
  "react/no-array-index-key": "error",
  "react/no-children-prop": 2,
  "react/no-danger-with-children": 2,
  "react/no-deprecated": 2,
  "react/no-direct-mutation-state": 2,
  "react/no-find-dom-node": 2,
  "react/no-is-mounted": 2,
  "react/no-render-return-value": 2,
  "react/no-string-refs": 2,
  "react/no-unescaped-entities": 2,
  "react/no-unknown-property": 2,
  "react/no-unsafe": 0,
  "react/prop-types": 2,
  "react/react-in-jsx-scope": 2,
  "react/require-render-return": 2,
  "react/self-closing-comp": "error",
  "react/void-dom-elements-no-children": "error",
}
`;
